File: credit.c
#include <cs50.h>
#include <stdio.h>

int main()
{
    // AMEX 15 digits, starts with 34 or 37
    // Mastercard 16 digits, starts with 51, 52, 53, 54, or 55
    // Visa 13 or 16 digits, starts with 4
    long n = get_long("Enter credit card number: \n");
    long num = n; // copy
    int sum = 0, prod, count = 0;
    // checksum
    while (n != 0)
    {
        int digit = n % 10;
        if (count % 2 == 0)
        {
            sum += digit;
        }
        else
        {
            prod = 2 * digit;
            while (prod != 0)
            {
                sum += prod % 10;
                prod /= 10;
            }
        }

        n /= 10;
        count++;
    }

    // identifying provider

    if (sum % 10 == 0)
    {
        long start = num;
        while (start >= 100)
        {
            start /= 10;
        }
        if ((start == 34 || start == 37) && (count == 15))
            printf("AMEX\n");
        else if (((start / 10) == 4) && (count == 13 || count == 16))
            printf("VISA\n");
        else if ((start >= 51 && start <= 55) && (count == 16))
            printf("MASTERCARD\n");
        else
            printf("INVALID\n");
    }
    else
        printf("INVALID\n");
}

File: credit.py
from cs50 import get_string


def card_check(card_num):
    sum = 0
    length = len(card_num)
    # adds digits in twice of alternating digits from second last number
    for i in range((length-2), -1, -2):
        d = int(card_num[i])*2
        for j in str(d):
            sum += int(j)
    # adds remaining digits
    for i in range((length-1), -1, -2):
        sum += int(card_num[i])
    # checks if sum ends with 0
    if sum % 10 == 0:
        return True
    else:
        return False


ccnum = get_string("Number: ")
# AMEX start with 34 or 37 - 15 digits
# MASTERCARD starts with 51,52,53,54 or 55 - 16 digits
# VISA starts with 4 - 13 or 16 digit numbers
length = len(ccnum)
if card_check(ccnum):
    #  chechikng for amex
    if length == 15 and (ccnum[0:2] == "34" or ccnum[0:2] == "37"):
        print("AMEX")
    #  checking for visa
    elif (length == 13 or length == 16) and ccnum[0] == "4":
        print("VISA")
    #  checking for mastercard
    elif length == 16 and (ccnum[0:2] in ["51", "52", "53", "54", "55"]):
        print("MASTERCARD")
    #  if neither than invalid
    else:
        print("INVALID")
# if card_check fails
else:
    print("INVALID")
